### DISCUSSION TITLE
Problem using SQLite in xplat project

### DISCUSSION THREAD

**Discussion by tsctsc6** [AvaloniaUI/Avalonia#16458] [Q&A] [answered]
I create a project from [SimpleToDoList](https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/CompleteApps/SimpleToDoList), which written in the official documentation.

I do two more things base on it.
* First, adapt it to the xplat project.
* Second, use SQLite to store ToDoItem.

Everything works well in Windows Platform. However, problem with Android Platform. When reopening the app, the previous record disappears.

The way I get the database file path is:
```
Path.Combine(AppContext.BaseDirectory, "MyToDoList.db");
```
In Android Platform, it is:
```
/data/user/0/com.CompanyName.SimpleToDoList/files
```
However, I can't find this path in Android device. What happend?

**Comment by maxkatz6** [comment]
> AppContext.BaseDirectory

I don't think you can realistically have write access to the app installation folder.

Instead, you should access to some special folders, see https://learn.microsoft.com/en-us/previous-versions/xamarin/android/platform/files/#working-with-internal-storage
Many samples use `SpecialFolder.Personal`, but you also can use `SpecialFolder.LocalApplicationData` for consistency with Windows.
Also, I recommend creating a subdirectory for your app. I.e. `Path.Combine(Environment.GetFolderPath(SpecialFolder.LocalApplicationData), "my-app-name", "MyToDoList.db")`.

**Reply by tsctsc6** [reply]
Unfortunately, it didn't work.

And I've set read and write permissions on Android.

`Environment.GetFolderPath(SpecialFolder.LocalApplicationData)` also is `/data/user/0/com.CompanyName.SimpleToDoList/files` on Android. Still, I can't find this path in Android device.

**Comment by tsctsc6** [comment]
OK...now I understand internal storage is inaccessible for ordinary user, only the app can access it.

And I create aother project, using `File.WriteAllText()` to write down something in `/data/user/0/com.CompanyName.SimpleToDoList/files/`. No problem.

But the problem at the beginning was not solved. Why my database can't record data? Could it be the database nuget package problem? I am using [sqlite-net-pcl](https://www.nuget.org/packages/sqlite-net-pcl/).

**Reply by timunie** [reply]
SqLite can be a headache in Android. I remember back in the xamarin days I had to get user permissions first. There should be some info about it for xamarin ir MAUI out in the www, should be applicable to Avalonia more or less.

**Reply by tsctsc6** [reply]
I followed [Microsoft's documentation](https://learn.microsoft.com/en-us/dotnet/maui/data-cloud/database-sqlite?view=net-maui-8.0) from the start, and the documentation doesn't say anything special about Android.

In that documentation, the database path is `FileSystem.AppDataDirectory` , but I can't call this function in Avalonia project.

**Comment by tsctsc6** [comment, marked as answer]
Geez, I realize this stems from a hidden mistake!

In App.axaml, function `OnFrameworkInitializationCompleted()` , when distinguishing between `IClassicDesktopStyleApplicationLifetime` and `ISingleViewApplicationLifetime` , creating `MainWindow` . This results in the data not being displayed in the GUI.

There's no problem now.

**Comment by tsctsc6** [comment]
By the way, what is `IClassicDesktopStyleApplicationLifetime` and `ISingleViewApplicationLifetime` ? When is the `ApplicationLifetime` `IClassicDesktopStyleApplicationLifetime` and when is it `ISingleViewApplicationLifetime` ?

**Reply by maxkatz6** [reply]
As interface name says, one is used with platforms running on a classic desktop, with proper windowing. And another with platforms where whole app has to work on a single view (mobile, browser).

**Reply by tsctsc6** [reply]
Got it!
